Extend dh/dt use in energy time derivative of NS to porous media

GiudGiud opened the pull request:
slowly making my way to fixing the VTB fhr inputs. 
I dont really want to break the VTB more hence why I am not disallowing the old option

GiudGiud committed:
Improve linear solve parameter name

GiudGiud committed:
Make enthalpy material in NSFV a functormaterial

GiudGiud committed:
Add an option to use the time derivative of the specific enthalpy
instead of the specific heat
in the time derivative of the porous NS energy equation refs #21245

GiudGiud committed:
Add a gas pwcns flow input using dh_dt

GiudGiud commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/gold/2d-transient-gas_out.csv:
```diff
@@ -0,0 +1,5 @@
+time,inlet-p,outlet-temp,outlet-u,solid-temp
+0,0,0,0,0
+1,8.777543485164,110.76196226955,-2.7731418339086,107.52928262687
```
side average value clearly not the way to get velocities :O

lindsayad replied:
wow. Are we going to keep this gold file?

GiudGiud replied:
yeah actually if you look at the exodus the gas moves weird on the first step
changed to VolFlowRate

lindsayad replied:
interesting

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1856428/) on 40b94bd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25922/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1856446/) on 40b94bd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25922 40b94b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.64%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92451</td>
      <td>92452</td>
      <td>+1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15509</td>
      <td>15508</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25922/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25922/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25922 40b94b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.77%</td>
      <td>84.76%</td>
      <td>-0.01%</td>
      <td>92.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12548</td>
      <td>12556</td>
      <td>+8</td>
      <td>37</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2255</td>
      <td>2258</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25922/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25922/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25922/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/2d-transient-gas.i:
```diff
@@ -0,0 +1,372 @@
+  []
+  [superficial_vel_y]
+    type = PINSFVSuperficialVelocityVariable
+    initial_condition = 1e-6
```
```suggestion
```

lindsayad commented on modules/navier_stokes/src/fvkernels/PINSFVEnergyTimeDerivative.C:
```diff
@@ -59,10 +72,16 @@ PINSFVEnergyTimeDerivative::computeQpResidual()
-      time_derivative +=
-          (*_rho_dot)(elem_arg, state) * _cp(elem_arg, state) * _var(elem_arg, state);
+    ADReal time_derivative;
+    if (_h_dot)
+      time_derivative = _rho(elem_arg, state) * (*_h_dot)(elem_arg, state);
+    else
+    {
+      time_derivative = _rho(elem_arg, state) * (*_cp)(elem_arg, state) * _var.dot(elem_arg, state);
+      if (_rho_dot)
```
In one branch you handle a time varying density, but in another you don't. It makes me wonder whether we should do it all in this PINSFV kernel or have a PWCNSFV extension. In `INSFVEnergyAdvection` you get no `rho_dot` handling

GiudGiud replied:
ah good catch.
This could probably just become PWCNSFV ?

lindsayad replied:
Yea probably could. Although we're kind of in a hard spot with names. A lot of INSFV objects are appropriate for WCNSFV simulations so maybe we should rename all of those objects since I would regard WCNSFV as covering INSFV but not the other way around(?)

lindsayad replied:
anyway I think what you did in [af75916](https://github.com/idaholab/moose/pull/25922/commits/af75916433a5ea7c09aa88f31965b62b2d948312) addresses this particular conversation

GiudGiud replied:
I m more ok with WCNSFV covering INSFV than INSFV covering WCNSFV and both are better than
INSFV sometimes covering WCNSFV sometimes not

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/2d-transient-gas.i:
```diff
@@ -0,0 +1,372 @@
+[]
+
+[AuxKernels]
+  [speed]
```
There's also the `PINSFVSpeedFunctorMaterial`

GiudGiud replied:
changed

GiudGiud replied:
changed

GiudGiud committed:
Address Alex's review
- add rho_dot h term to the time derivative for weakly-compressible
- simplify input and use best-practice PP & speed functor

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/2d-transient-gas.i:
```diff
@@ -0,0 +1,374 @@
+    type = MooseVariableFVReal
+    initial_condition = 0.5
+  []
+  [velocity_norm]
+    type = MooseVariableFVReal
+  []
```
```suggestion
```

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/pwcns/channel-flow/gold/2d-transient-gas_out.csv:
```diff
@@ -0,0 +1,5 @@
+time,inlet-p,outlet-temp,outlet-u,solid-temp
+0,0,0,0,0
+1,8.6557148893077,106.39295004178,-2.9848928309229,104.9552224657
```
Still weird with the PP type change. Is there actually an inflow at this first time step or is this still a postprocessing artifact?

GiudGiud replied:
my bet is inflow during the transient. Lemme send you a gif

lindsayad replied:
this is already discussed in another conversation

GiudGiud committed:
Remove deprecated name for INSFVEnthalpyMaterial

GiudGiud commented:
10th time s the charm lol
